### PR TITLE
ggml alloc: Fix for null dereference on alloc failure

### PR DESCRIPTION
**Pull request by ptsochantaris**
If `alloc_tensor_range` is called and either/or:
- The `buffers` argument is NULL
- The `ggml_tallocr_new_from_buffer` call inside this function fails to allocate

The code will end up trying to free a NULL `buffers` pointer which, at least on macOS, causes a crash (or worse, corruption)

This PR adds checks for `buffers` being NULL, or for some reason `buffers` points to NULL. This change will allow the llama context init to fail (as expected, since there was not enough memory in this path) rather than crash / corrupt state.

(Note: This breakage will happen in release builds. In debug builds the Metal framework will throw an exception during the allocation if there is not enough memory, so the code will not reach this point, but in release builds it will keep going and break)

Picture of the call stack in case it helps:

Screenshot 2024-01-29 at 21 16 43

To reproduce: Load a model that is much bigger than the system memory and allocate all the layers to the GPU on an Apple Silicon Mac.

### CODE REVIEW

File: ggml-alloc.c [hunk 1]
```diff
@@ -788,10 +788,12 @@ static bool alloc_tensor_range(struct ggml_context * ctx,
 #ifndef NDEBUG
         fprintf(stderr, "%s: failed to allocate %s buffer of size %zu\n", __func__, ggml_backend_buft_name(buft), size);
 #endif
-        for (size_t i = 0; i < *n_buffers; i++) {
-            ggml_backend_buffer_free(*buffers[i]);
+        if (buffers && *buffers) {
```
Comment (slaren) [inline ggml-alloc.c:791]:
```suggestion
        if (*buffers) {
```

File: ggml-alloc.c [hunk 2]
```diff
@@ -843,10 +845,12 @@ ggml_backend_buffer_t ggml_backend_alloc_ctx_tensors_from_buft(struct ggml_conte
                     __func__, t->name,
                     ggml_backend_buft_name(buft),
                     this_size, max_size);
-            for (size_t i = 0; i < n_buffers; i++) {
-                ggml_backend_buffer_free(buffers[i]);
+            if(buffers && *buffers) {
```
Comment (slaren) [inline ggml-alloc.c:848]:
```suggestion
            if (buffers) {
```

**Comment by slaren** [comment]
I am not sure if I understand this issue. In `alloc_tensor_range`, `buffers` will never be `NULL` (it points to a local in `ggml_backend_alloc_ctx_tensors_from_buft`), and if the allocation fails `*n_buffers` will be zero so it will never be dereferenced.

**Comment by ptsochantaris** [comment]
Apologies if I'm missing something in the code structure, but from what I can tell in this method, `buffers` starts off as NULL on line 829, and in some cases will reach one of the calls to `alloc_tensor_range` in line 855 and/or 867 with a NULL value.

As an example, if buffer is NULL and buffers is also NULL (perhaps on the first call through the loop?) then there is a null pointer free on lines 794. 

The second code path that can lead to the same this is if `this_size > max_size` is true on the first run through the loop (where `buffers` is NULL), and possibly others.

I hope this makes some sense, but again, apologies if I'm missing something here, I'm not too familiar with this section of code. I just noticed the explosion on my project and thought this may help catch this corner case in the library.

Screenshot 2024-01-29 at 21 52 16

**Comment by slaren** [comment]
I think I see the issue, the `free(buffers)` in `alloc_tensor_range` should be `free(*buffers)` instead. Can you check if that change alone fixes the issue?

**Comment by ptsochantaris** [comment]
(BTW worth mentioning that one alternative is to check for `n_buffers == 0` instead of the `buffers` pointer being NULL, however that would still be vulnerable to a corner case where somehow because of some other bug `n_buffers` was zero but the `buffers` pointer needed freeing anyway.)

**Comment by ptsochantaris** [comment]
Re: Freeing a null pointer - aha! Let's me check this now to see if it fixes it.

**Comment by ptsochantaris** [comment]
Yes, that change does indeed do the trick. My solution was way over engineered, in this case :D I'll update the PR shortly, good shout 👍

**Comment by ptsochantaris** [comment]
I've updated the three `free(buffers)` instances in the file, I assume they all needed changing - but please sanity check me :)

**Comment by slaren** [comment]
Only use `free(*buffers)` in `alloc_tensor_range`, but leave the frees unchanged in `ggml_backend_alloc_ctx_tensors_from_buft`. The issue is that in `alloc_tensor_range`, `buffers` is a pointer to a pointer, but in `ggml_backend_alloc_ctx_tensors_from_buft` it is just a pointer.

**Comment by ptsochantaris** [comment]
Ah, indeed, thanks, that one was a double pointer. Sorry for being slow. Will revert to only adding the one `*`. #minimalism

**Review by slaren — APPROVED** [review]

**Comment by slaren** [comment]
Thank you for finding and fixing the issue!

**Pull request merged**